src/plugins/mavlink_ftp/mavlink_ftp_impl.cpp

```diff
@@ -0,0 +1,1121 @@
+        _parent->get_own_component_id(),
+        &_last_command,
+        _network_id,
+        _parent->get_system_id(),
+        _get_target_component_id(),
+        raw_payload);
+    _parent->send_message(_last_command);
+
+    _reset_timer();
+    if (!_last_command_timer_running) {
+        _last_command_timer_running = true;
+        _last_command_timer = std::async(std::launch::async, [&]() {
+            while (true) {
+                _timer_mutex.lock();
+                if (!_last_command_timer_running) {
+                    _timer_mutex.unlock();
+                    return;
+                }
+                auto now = std::chrono::steady_clock::now();
+                auto time_diff = std::chrono::duration_cast<std::chrono::milliseconds>(
+                    now - _last_command_start_time);
+                if (time_diff.count() > _last_command_timeout) {
+                    if (_last_command_retries >= _max_last_command_retries) {
+                        LogErr() << "Response timeout " << _curr_op;
+                        _last_command_timer_running = false;
+                        _session_result = ServerResult::ERR_TIMEOUT;
+                        _session_valid = false;
+                        _timer_mutex.unlock();
+                        _process_nak(ServerResult::ERR_TIMEOUT);
+                        return;
+                    }
+                    _last_command_retries++;
+                    _last_command_start_time = now;
+                    LogWarn() << "Response timeout. Retry: " << _last_command_retries;
+                    _parent->send_message(_last_command);
+                }
+                _timer_mutex.unlock();
+                std::this_thread::sleep_for(std::chrono::milliseconds{10});
+            }
+        });
+    }
+}
+
+void MavlinkFTPImpl::_reset_timer()
+{
+    _last_command_start_time = std::chrono::steady_clock::now();
+    _last_command_retries = 0;
+}
+
+void MavlinkFTPImpl::_stop_timer()
+{
+    {
+        std::lock_guard<std::mutex> lock(_timer_mutex);
+        if (!_last_command_timer_running) {
+            return;
+        }
+        _last_command_timer_running = false;
+    }
+    _last_command_timer.wait();
+}
+
+void MavlinkFTPImpl::process_mavlink_ftp_message(const mavlink_message_t& msg)
+{
+    if (msg.msgid != MAVLINK_MSG_ID_FILE_TRANSFER_PROTOCOL) {
+        return;
+    }
+
+    bool stream_send = false;
+    mavlink_file_transfer_protocol_t ftp_req;
+    mavlink_msg_file_transfer_protocol_decode(&msg, &ftp_req);
+
+    if (ftp_req.target_component != 0 && ftp_req.target_component != get_our_compid()) {
+        return;
+    }
+
+    PayloadHeader* payload = reinterpret_cast<PayloadHeader*>(&ftp_req.payload[0]);
+
+    ServerResult errorCode = ServerResult::SUCCESS;
+
+    // basic sanity checks; must validate length before use
+    if (payload->size > max_data_length) {
+        errorCode = ServerResult::ERR_INVALID_DATA_SIZE;
+        goto out;
```

Comment:
Any chance you can use `return` early instead of `goto`? There is only this one, so an `if` would probably do it.

Reply:
Changed it. (The original code was taken from Firmware)